### PR TITLE
Relative paths (issue #622)

### PR DESCRIPTION
**Pull request by lamarpavel**
As discussed in [issue622](https://github.com/The-Compiler/qutebrowser/issues/622) this PR adds the option to specify paths to (some) config files relative to the config-dir of qutebrowser.

One test in tests/config/test_configtypes.py needed to be modified.

### CODE REVIEW

File: qutebrowser/config/configtypes.py [hunk 1]
```diff
@@ -798,6 +799,17 @@ class File(BaseType):
 
     typestr = 'file'
 
+    def transform(self, value):
+        if not value:
+            return None
+        value = os.path.expanduser(value)
+        if not os.path.isabs(value):
+            if standarddir.config():
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: I'd move this check to `validate` - right now I think your code will raise an exception when started with `-c ''` (no configuration), because in `validate` you don't check for `None`. And when you check for there, there's no need to check in `transform` as well.

Also, I'd be glad if you could add a test for that scenario, though if you get stuck there let me know and I'll add it. It would probably look something like this (untested):

```python
def test_validate_rel_config_none(self, os_path, monkeypatch):
    """Test with a relative path and standarddir.config() returning None."""
    monkeypatch.setattr('qutebrowser.config.configtypes.standarddir.config', lambda: None)
    with pytest.raises(configexc.ValidationError):
        self.t.validate('foobar')
```

File: qutebrowser/config/configtypes.py [hunk 2]
```diff
@@ -806,6 +818,10 @@ def validate(self, value):
                 raise configexc.ValidationError(value, "may not be empty!")
         value = os.path.expanduser(value)
         try:
+            if not os.path.isabs(value):
+                abspath = os.path.join(standarddir.config(), value)
+                if os.path.isfile(abspath):
+                    return
             if not os.path.isfile(value):
                 raise configexc.ValidationError(value, "must be a valid file!")
             if not os.path.isabs(value):
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: This seems like dead code after your addition above

File: qutebrowser/config/configtypes.py [hunk 3]
```diff
@@ -806,6 +818,10 @@ def validate(self, value):
                 raise configexc.ValidationError(value, "may not be empty!")
         value = os.path.expanduser(value)
         try:
+            if not os.path.isabs(value):
+                abspath = os.path.join(standarddir.config(), value)
+                if os.path.isfile(abspath):
+                    return
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: Is there a reason you set and check `abspath` separately instead of just doing `value = os.path.join(...)` and then using the existing check below?

File: qutebrowser/config/configtypes.py [hunk 4]
```diff
@@ -1151,6 +1162,17 @@ class UserStyleSheet(File):
     def __init__(self):
         super().__init__(none_ok=True)
 
+    def transform(self, value):
+        if not value:
+            return None
+        path = super().transform(value)
+        path = os.path.expandvars(path)
+        if os.path.isabs(path):
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: Shouldn't this rather be `os.path.exists` now?

File: qutebrowser/config/configtypes.py [hunk 5]
```diff
@@ -1151,6 +1162,17 @@ class UserStyleSheet(File):
     def __init__(self):
         super().__init__(none_ok=True)
 
+    def transform(self, value):
+        if not value:
+            return None
+        path = super().transform(value)
+        path = os.path.expandvars(path)
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: Before `os.path.expanduser` was also called which isn't anymore - but actually both `expanduser` and `expandvars` should be called in `File.transform` IMHO. I'm not sure why `expandvars` isn't there.

File: qutebrowser/config/configtypes.py [hunk 6]
```diff
@@ -1160,31 +1182,21 @@ def validate(self, value):
...
+                    # FIXME We just try if it is encodable, maybe we should
+                    # validate CSS?
+                    # https://github.com/The-Compiler/qutebrowser/issues/115
+                    try:
+                        value.encode('utf-8')
+                    except UnicodeEncodeError as e:
+                        raise configexc.ValidationError(value, str(e))
+                    return
+            except UnicodeEncodeError as e:
+                raise configexc.ValidationError(value, e)
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: Keeping this here is actually good - I just didn't think about this in the example I did give you. Good we have tests! :smile:

File: tests/config/test_configtypes.py [hunk 7]
```diff
@@ -1378,8 +1378,7 @@ def test_validate_exists_not_abs(self, os_path):
         os_path.expanduser.side_effect = lambda x: x
         os_path.isfile.return_value = True
         os_path.isabs.return_value = False
-        with pytest.raises(configexc.ValidationError):
-            self.t.validate('foobar')
+        self.t.validate('foobar')
```
Comment (The-Compiler) [inline tests/config/test_configtypes.py]: Changing this test makes sense, as relative paths work now. Could you please also rename the test to `..._rel` instead `..._not_abs` (and the comment to say `relative` instead of `not absolute`) as this is the "normal case" now?

Also, I think below you could add `os_path.isfile.assert_called_once_with(...)` and the same for `os_path.isabs` to make sure the paths get expanded correctly with `standarddir.config()`.

File: qutebrowser/config/configtypes.py (hunk 3 above)
Comment (lamarpavel) [inline qutebrowser/config/configtypes.py]: Now that you mention it, the first check seems irrelevant. For the creation of ```abspath``` there is no other reason that coding style. I initially wanted to write it in one line like this:
```
if os.path.isfile(os.path.join(standarddir.config(), value)):
    return
```
But looking at the other code it seemed to me that a more verbose style was preferred (eg. two separate lines for expanduser and expandvars instead of ```os.path.expanduser(os.path.expandvars(...))```)

File: qutebrowser/config/configtypes.py (hunk 5 above)
Comment (lamarpavel) [inline qutebrowser/config/configtypes.py]: I initially put both expanduser and expandvars in File.transform and removed them from UserStyleSheet.transform, but that triggered
```FAIL tests/config/test_configtypes.py::TestFile::()::test_transform```
and I couldn't get my head around why. So I left expandvars in UserStyleSheet.transform for now. I must admit that the test cases are still somewhat over my head.

Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: Those tests currently use Python's [unittest.mock](https://docs.python.org/3/library/unittest.mock.html) library, and `os_path` is the mock object which replaces the real `os.path.<function>` calls in the `conftypes.py` module.

One rather surprising thing about mocks is that any call to them return *another* mock object if you didn't configure them to do something else via `the_mock.return_value = ...` or `the_mock.side_effect = ...`.

At the beginning of the test, we do `os_path.expanduser.side_effect = lambda x: x.replace('~', '/home/foo')`. This assigns an anonymous function (a [lambda](https://docs.python.org/3/reference/expressions.html#lambda)) to `expanduser` which returns the given string with `~` replaced by `/home/foo`.

However, we don't do the same with `os_path.expandvars` but call `expandvars` from our code - which causes `os_path.expandvars(...)` to return another mock object (a `MagicMock`), which fails our test:

```
>       assert self.t.transform('~/foobar') == '/home/foo/foobar'
E       assert <MagicMock name='path.expandvars()' id='139799983029048'> == '/home/foo/foobar'
```

The solution here is to provide some stub for `expandvars`. I got it working with this diff:

```diff
diff --git a/qutebrowser/config/configtypes.py b/qutebrowser/config/configtypes.py
index 4b0b5b7..2ecb303 100644
--- a/qutebrowser/config/configtypes.py
+++ b/qutebrowser/config/configtypes.py
@@ -803,6 +803,7 @@ class File(BaseType):
         if not value:
             return None
         value = os.path.expanduser(value)
+        value = os.path.expandvars(value)
         if not os.path.isabs(value):
             if standarddir.config():
                 abspath = os.path.join(standarddir.config(), value)
@@ -1166,7 +1167,6 @@ class UserStyleSheet(File):
         if not value:
             return None
         path = super().transform(value)
-        path = os.path.expandvars(path)
         if os.path.isabs(path):
             return QUrl.fromLocalFile(path)
         else:
diff --git a/tests/config/test_configtypes.py b/tests/config/test_configtypes.py
index c515c6c..0f7b355 100644
--- a/tests/config/test_configtypes.py
+++ b/tests/config/test_configtypes.py
@@ -1398,6 +1398,7 @@ class TestFile:
     def test_transform(self, os_path):
         """Test transform."""
         os_path.expanduser.side_effect = lambda x: x.replace('~', '/home/foo')
+        os_path.expandvars.side_effect = lambda x: x
         assert self.t.transform('~/foobar') == '/home/foo/foobar'
         os_path.expanduser.assert_called_once_with('~/foobar')
```

But looking at the tests again now, maybe I should rewrite them to use real files in a temporary directory instead - I agree mocks can be hard to grasp.

File: qutebrowser/config/configtypes.py (hunk 4 above)
Comment (lamarpavel) [inline qutebrowser/config/configtypes.py]: Testing with ```isfile``` or ```exists``` results in two failed tests
```
FAIL TestUserStyleSheet::()::test_transform_file
FAIL TestUserStyleSheet::()::test_transform_file_expandvars
```
If I interpret that right the first one fails because the file **file:///foo/bar** doesn't actually exist and we thus reach the else. In that case the test probably needs to change, but I'm not sure how. Should it create a file first and then expect to find it?

Same goes for the second failed test, I think.

Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: You could use the `os_path` mock to fake `os.path.exists` to return True:

```python
def test_transform_file(self, os_path):
    """Test transform with a filename."""
    os_path.exists.return_value = True
    ...
```

(see above for an explanation of the mocks)

File: qutebrowser/config/configtypes.py (hunk 5 above)
Comment (lamarpavel) [inline qutebrowser/config/configtypes.py]: Thanks for the thorough explanation, this helps a lot. I will change the tests for now to accept the corrected code.

File: tests/config/test_configtypes.py (hunk 7 above)
Comment (lamarpavel) [inline tests/config/test_configtypes.py]: The test for assert_called_once_with() is indeed justified, standarddir.config() actually returns None during that test. I'll try to fix this and the remaining issues tomorrow, I'm out of time for today.

File: qutebrowser/config/configtypes.py (hunk 1 above)
Comment (lamarpavel) [inline qutebrowser/config/configtypes.py]: I included the check in validate, updated the corresponding test and added the new test (had to add a mock for os_path.isabs), however when removing the test for the return of ```standarddir.config()``` ```test_transform_base64``` of ```TestUserStyleSheet``` fails.

The test isn't calling validate before transform, but even if it would, it would catch the ValidationError and continue even though standarddir.config() returns None, at which point it will be trigger that same trap in File.transform. I think the easiest solution is to just leave the test for the return of standarddir.config() in File.transform.

File: qutebrowser/config/configtypes.py [hunk 8]
```diff
@@ -798,6 +799,18 @@ class File(BaseType):
 
     typestr = 'file'
 
+    def transform(self, value):
+        if not value:
+            return None
+        value = os.path.expanduser(value)
+        value = os.path.expandvars(value)
+        if not os.path.isabs(value):
+            if standarddir.config():
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: Please explicitely check for `... is not None` here - in this case it wouldn't matter as it never should be anything else falsey (i.e. it'd never return an empty string), but it still makes it more clear this is expected to be a string or `None`.

File: qutebrowser/config/configtypes.py [hunk 9]
```diff
@@ -806,19 +819,17 @@ def validate(self, value):
                 raise configexc.ValidationError(value, "may not be empty!")
         value = os.path.expanduser(value)
         try:
+            if not os.path.isabs(value):
+                cfgdir = standarddir.config()
+                if cfgdir and os.path.isfile(os.path.join(cfgdir, value)):
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: Please also check `if cfgdir is not None and ...` here (see above)

File: qutebrowser/config/configtypes.py [hunk 10]
```diff
@@ -806,19 +819,17 @@ def validate(self, value):
...
+                    return
+                raise configexc.ValidationError(value,
+                                                "must be an absolute path!")
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: I think the error message(s) should be a bit more fine-grained here, otherwise it might be confusing.

I'd say:

- `if cfgdir is None:`, raise a `ValidationError` with something like "must be an absolute path if no config directory is used!"
- If it's not a file, raise a `ValidationError` with something like "must be an existing file relative to the config directory"

File: qutebrowser/config/configtypes.py [hunk 11]
```diff
@@ -806,19 +819,17 @@ def validate(self, value):
...
             if not os.path.isfile(value):
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: Even if you'd always return or raise above, this should probably be an `elif`.

File: qutebrowser/config/configtypes.py [hunk 12]
```diff
@@ -1160,31 +1181,21 @@ def validate(self, value):
...
+                if not os.path.isabs(value):
+                    # probably a CSS, so we don't handle it as filename.
+                    # FIXME We just try if it is encodable, maybe we should
+                    # validate CSS?
+                    # https://github.com/The-Compiler/qutebrowser/issues/115
+                    try:
+                        value.encode('utf-8')
+                    except UnicodeEncodeError as e:
+                        raise configexc.ValidationError(value, str(e))
+                    return
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: I didn't see this before (in my code), but this should be simplified to a simple `value.encode('utf-8')` because `UnicodeEncodeError` is handled below already anyways, and it's the end of the function anyways.

File: qutebrowser/config/configtypes.py (hunk 6 above)
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: This should probably be changed to `str(e)` like above - my bad!

**Comment by The-Compiler** [comment]
I left some more comments - sorry for the delay! Other than that it looks good - and thanks for the good tests!

**Comment by lamarpavel** [comment]
There you go.

P.S.: I merged in upstream changes to make sure all tests still work, which resulted in about 80 additional commits being part of this branch. I didn't think about that beforehand and I'm not sure if that's good practice for PRs, should I avoid that and let you handle the merge in the future?

File: qutebrowser/config/configtypes.py [hunk 13]
```diff
@@ -798,6 +799,18 @@ class File(BaseType):
...
 
+    def transform(self, value):
+        if not value:
+            return None
+        value = os.path.expanduser(value)
+        value = os.path.expandvars(value)
+        if not os.path.isabs(value):
+            if standarddir.config() is not None:
+                abspath = os.path.join(standarddir.config(), value)
+                if os.path.isfile(abspath):
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: I'm confused about this check - isn't the same thing already checked in `validate`? If so, no need to check it here again.

File: qutebrowser/config/configtypes.py [hunk 14]
```diff
@@ -806,19 +819,21 @@ def validate(self, value):
...
-                raise configexc.ValidationError(value, "must be a valid file!")
             if not os.path.isabs(value):
+                cfgdir = standarddir.config()
+                if cfgdir is not None and os.path.isfile(
+                        os.path.join(cfgdir, value)):
+                    return
                 raise configexc.ValidationError(
-                    value, "must be an absolute path!")
+                    value, "must be an absolute path when not using a config "
+                    "directory!")
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: This would also be raised if `cfgdir` is `None` but the file doesn't exist - I'd say something like this would make more sense:

```python
if cfgdir is None:
    raise configexc.ValidationError(
                    value, "must be an absolute path when not using a config "
                    "directory!")
if not os.path.isfile(os.path.join(cfgdir, value)):
    raise configexc.ValidationError(
                    value, "must be a valid path relative to the config "
                    "directory!")
```

File: qutebrowser/config/configtypes.py [hunk 15]
```diff
@@ -806,19 +819,21 @@ def validate(self, value):
...
+            elif not os.path.isfile(value):
+                raise configexc.ValidationError(
+                    value, "must be a valid path relative to the config "
+                    "directory!")
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: This would be raised if the user gives you an *absolute* path to something else than an existing file, so I think this should just say `"must be a valid file!"` or so.

**Comment by The-Compiler** [comment]
And another round :wink:

After that I'm quite sure I can merge this - I'm sorry if this is taking a bit longer than anticipated, but we're close :smile:

File: qutebrowser/config/configtypes.py (hunk 13 above)
Comment (lamarpavel) [inline qutebrowser/config/configtypes.py]: Oh yes. I think I've been a bit too caught up in trying to fix what you point out that I haven't looked at the bigger picture. I'll take a bit more time to review it this time, sorry. Also, I've got exams coming up, so it may take a bit longer.

**Comment by lamarpavel** [comment]
Here you go, I even threw in a bonus in the form of a new test (:

**Comment by The-Compiler** [comment]
Great, thanks! Looks good now from a quick look, will take another close look and (hopefully) merge later today.

**Comment by The-Compiler** [comment]
Merged, thank you! The only thing which was missing was [adjusting the docs](https://github.com/The-Compiler/qutebrowser/commit/5a2d909607e05a69d4500e022de73f20ce13886c) but since it's a small change I did that myself.

Regarding your question about merging master: It doesn't really matter much - myself I try to avoid merging master into feature-branches, but it's okay.

Another possibility if you just want to run the tests would be to create a temporary branch with relapath merged:

```
$ git checkout -b relapaths-merged master
$ git merge relapaths
$ tox
$ ...
$ git checkout relapaths
$ git branch -D relapaths-merged
```
